Switch MetaDataStateFormat to Lucene directory abstraction

andrershov opened the pull request:
Switching to Lucene directory abstraction for MetaDataStateFormat to be able to inject failure during writing state to disk.
Adding new tests to MetaDataStateFormat to inject these failures.

elasticmachine commented:
Pinging @elastic/es-distributed

ywelsch commented on server/src/test/java/org/elasticsearch/gateway/MetaDataStateFormatTests.java:
```diff
@@ -329,8 +420,31 @@ private MetaData randomMeta() throws IOException {
+         * @param failureMethods - one of these method calls will fail on Directory object if injectFailure is set to true.
+         *                       Method that will fail is randomly selected on each newDirectory method call.
+         */
+        Format(String prefix, AtomicBoolean injectFailure, String... failureMethods) {
```
instead of injecting the AtomicBoolean, it might be nicer to just create it internally, and access it on the `Format` object

andrershov replied:
++

ywelsch commented on server/src/test/java/org/elasticsearch/gateway/MetaDataStateFormatTests.java:
```diff
@@ -291,6 +292,96 @@ public void testLoadState() throws IOException {
+        DummyState state = new DummyState(randomRealisticUnicodeOfCodepointLengthBetween(1, 100), randomInt(), randomLong(), randomDouble(), randomBoolean());
+        format.write(state, paths);
+        assertEquals(state, format.loadLatestState(logger, NamedXContentRegistry.EMPTY, paths));
+        return state;
```
can we check that all other files have been properly cleaned up?

andrershov replied:
++

bleskes commented on server/src/main/java/org/elasticsearch/gateway/MetaDataStateFormat.java:
```diff
@@ -207,20 +225,15 @@ protected Directory newDirectory(Path dir) throws IOException {
+        for (Path location: locations){
+            Path stateLocation = location.resolve(STATE_DIR_NAME);
+            try (Directory stateDir = newDirectory(stateLocation)){
+                for (String file : stateDir.listAll()){
```
Tip, just in case: you can let intellij auto format a selection (and not the entire file) but selecting it and choosing reformat code from the Code menu (or using whatever keyboard shortcut you have).

andrershov replied:
Sorry about that, I know about format option and use it, but sometimes forget to apply it. Unfortunately, there is no good way for IDEA to do it safe and automatically for you

andrershov commented on server/src/main/java/org/elasticsearch/gateway/MetaDataStateFormat.java:
```diff
@@ -207,20 +225,15 @@ protected Directory newDirectory(Path dir) throws IOException {
-                    Files.deleteIfExists(stateFile);
-                    logger.trace("cleanupOldFiles: cleaned up {}", stateFile);
+    private void cleanupOldFiles(final String currentStateFile, Path[] locations) throws IOException {
+        for (Path location: locations){
```
++

andrershov commented on server/src/main/java/org/elasticsearch/gateway/MetaDataStateFormat.java:
```diff
@@ -101,60 +155,24 @@ public final void write(final T state, final Path... locations) throws IOExcepti
-                    Files.deleteIfExists(tmpPath);
-                    logger.trace("cleaned up {}", tmpPath);
-                }
+            for (Path extraLocation : locations) {
```
Good catch!

andrershov commented on server/src/main/java/org/elasticsearch/gateway/MetaDataStateFormat.java:
```diff
@@ -101,60 +155,24 @@ public final void write(final T state, final Path... locations) throws IOExcepti
+        final Path firstStateLocation = locations[0].resolve(STATE_DIR_NAME);
+        try (Directory stateDir = newDirectory(firstStateLocation)) {
+            writeStateToFirstLocation(state, stateDir, fileName, tmpFileName);
+            logger.trace("written state to {}", firstStateLocation.resolve(fileName));
```
It's done in this way because there I did not want to pass firstStateLocation path to writeStateToFirstLocation method just for logging purposes. Unfortunately, there is no method like resolve in Lucene directory. But fixed anyway

andrershov commented on server/src/main/java/org/elasticsearch/gateway/MetaDataStateFormat.java:
```diff
@@ -101,60 +155,24 @@ public final void write(final T state, final Path... locations) throws IOExcepti
+            for (Path extraLocation : locations) {
+                final Path extraStateLocation = extraLocation.resolve(STATE_DIR_NAME);
+                copyStateToExtraLocation(stateDir, extraStateLocation, fileName, tmpFileName);
+                logger.trace("copied state to {}", extraStateLocation.resolve(fileName));
```
Here it was done to make symmetrical with writeStateToFirstLocation, fixed

andrershov commented on server/src/main/java/org/elasticsearch/gateway/MetaDataStateFormat.java:
```diff
@@ -207,20 +225,15 @@ protected Directory newDirectory(Path dir) throws IOException {
-        };
-        // now clean up the old files
-        for (Path dataLocation : locations) {
-            logger.trace("cleanupOldFiles: cleaning up {}", dataLocation);
```
Same reason as writeStateToFirstLocation, fixed

ywelsch replied:
This looks to be still missing.

andrershov commented on server/src/main/java/org/elasticsearch/gateway/MetaDataStateFormat.java:
```diff
@@ -101,60 +155,24 @@ public final void write(final T state, final Path... locations) throws IOExcepti
-                    logger.trace("copied state to {}", finalPath);
-                } finally {
-                    Files.deleteIfExists(tmpPath);
-                    logger.trace("cleaned up {}", tmpPath);
```
Same reason as writeStateToFirstLocation, fixed

andrershov commented on server/src/main/java/org/elasticsearch/gateway/MetaDataStateFormat.java:
```diff
@@ -101,60 +155,24 @@ public final void write(final T state, final Path... locations) throws IOExcepti
             }
-        } finally {
-            Files.deleteIfExists(tmpStatePath);
-            logger.trace("cleaned up {}", tmpStatePath);
```
++

andrershov commented on server/src/test/java/org/elasticsearch/gateway/MetaDataStateFormatTests.java:
```diff
@@ -291,6 +292,96 @@ public void testLoadState() throws IOException {
+        DummyState initialState = writeAndReadStateSuccessfully(format, injectFailures, path);
+        possibleStates.add(initialState);
+
+        for (int i=0; i<randomIntBetween(1,5); i++) {
```
++

andrershov commented on server/src/test/java/org/elasticsearch/gateway/MetaDataStateFormatTests.java:
```diff
@@ -344,8 +458,22 @@ public DummyState fromXContent(XContentParser parser) throws IOException {
+                    public void eval(MockDirectoryWrapper dir) throws IOException {
+                        for (StackTraceElement e : Thread.currentThread().getStackTrace()) {
+                                if (failMethod.equals(e.getMethodName())) {
+                                    throw new MockDirectoryWrapper.FakeIOException();
```
Yes, it will fail on the first method call which name equals to one of randomly selected from failure methods.
For example, if you look at "testFailWriteAndReadPreviousState" you will find that failure methods are the following: "createOutput", "sync", "rename". If we run the sufficient number of iterations of the test, we can be sure that we test failures at all possible stages.

andrershov commented:
@ywelsch thanks for the review! I've made necessary changes in afa9a3b24b96db0719541153f7fafdae26311a66. Please let me know if something else is needed.
By injecting failures during writes did you mean, that we also should test for failures when writing bytes of the temporary file to disk? If yes, I've addressed it.

ywelsch commented on server/src/main/java/org/elasticsearch/gateway/MetaDataStateFormat.java:
```diff
@@ -207,20 +225,15 @@ protected Directory newDirectory(Path dir) throws IOException {
+                for (String file : stateDir.listAll()) {
+                    if (file.startsWith(prefix) && file.equals(currentStateFile) == false) {
+                        deleteTempFile(stateLocation, stateDir, file);
+                        logger.trace("cleanupOldFiles: cleaned up {}", stateLocation.resolve(file));
```
this is already logged part of `deleteTempFile`?

andrershov replied:
Fixed, also renamed deleteTempFile method, because when used in cleanup it's not temp file, but previous state

andrershov commented:
@ywelsch I've fixed logging issues and added the random test. Random test discovered the bug, that we don't take temp files into account when looking up max state id. I've fixed that.
Also, I've refactored loadLatestState method to remove code duplication.
May I ask you to kindly perform one more code review pass?

ywelsch commented on server/src/main/java/org/elasticsearch/gateway/MetaDataStateFormat.java:
```diff
@@ -245,6 +267,23 @@ long findMaxStateId(final String prefix, Path... locations) throws IOException {
+        for (Path dataLocation : locations) {
+            final Path stateFilePath = dataLocation.resolve(STATE_DIR_NAME).resolve(fileName);
+            if (Files.exists(stateFilePath)) {
+                files.add(stateFilePath);
```
Let's add trace logging to this (like it was before):

```
logger.trace("found state file: {}", stateFilePath);
```

ywelsch commented on server/src/test/java/org/elasticsearch/gateway/MetaDataStateFormatTests.java:
```diff
@@ -291,6 +291,113 @@ public void testLoadState() throws IOException {
+    }
+
+    public void testFailRandomlyAndReadAnyState() throws IOException {
+        Path paths[] = new Path[randomIntBetween(2, 5)];
```
let's also consider the typical use case of a single data path :)

andrershov commented:
@ywelsch thank you for another round, I've made changes in this commit 0f33a39911dfc21e9f42ebd840212eda336519c1

andrershov commented:
run gradle build tests please

andrershov commented:
run gradle build tests please

andrershov commented:
run gradle build tests again
